File: atmibroker-corba/src/main/cpp/ConnectionImpl.cxx
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat, Inc., and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
#ifdef TAO_COMP
#include <orbsvcs/CosNamingS.h>
#endif

#include <string.h>
#include "ConnectionImpl.h"
#include "SessionImpl.h"
#include "OrbManagement.h"
#include "txClient.h"
#include "AtmiBrokerPoaFac.h"
#include "EndpointQueue.h"

log4cxx::LoggerPtr CorbaConnectionImpl::logger(log4cxx::Logger::getLogger(
		"CorbaConnectionImpl"));

CorbaConnectionImpl::CorbaConnectionImpl(char* connectionName) {
	LOG4CXX_DEBUG(logger, (char*) "constructor");
	this->connection = (CORBA_CONNECTION *) start_tx_orb(connectionName);
}

CorbaConnectionImpl::~CorbaConnectionImpl() {
	LOG4CXX_DEBUG(logger, (char*) "destructor");
	shutdownBindings(this->connection);
}

Session* CorbaConnectionImpl::createSession(int id, char * serviceName) {
	LOG4CXX_DEBUG(logger, (char*) "createSession: " + id);
	sessionMap[id] = new CorbaSessionImpl(this->connection, id, serviceName);
	return sessionMap[id];
}

Session* CorbaConnectionImpl::createSession(int id, const char* temporaryQueueName) {
	LOG4CXX_DEBUG(logger, (char*) "createSession");
	return new CorbaSessionImpl(this->connection, id, temporaryQueueName);
}

Session* CorbaConnectionImpl::getSession(int id) {
	return sessionMap[id];
}

void CorbaConnectionImpl::closeSession(int id) {
	if (sessionMap[id]) {
		delete sessionMap[id];
		sessionMap[id] = NULL;
	}
}

Destination* CorbaConnectionImpl::createDestination(char* serviceName) {
	// create Poa for Service Queue
	AtmiBrokerPoaFac* poaFactory = this->connection->poaFactory;
	PortableServer::POA_ptr aFactoryPoaPtr = poaFactory->createServicePoa(
			this->connection->orbRef, serviceName, this->connection->root_poa,
			this->connection->root_poa_manager);
	LOG4CXX_DEBUG(logger, (char*) "created create_service_factory_poa: "
			<< aFactoryPoaPtr);

	return new CorbaEndpointQueue(this->connection, aFactoryPoaPtr, serviceName);
}

void CorbaConnectionImpl::destroyDestination(Destination* destination) {
	char* serviceName = strdup(destination->getName());
	CosNaming::Name * name = this->connection->default_ctx->to_name(
			destination->getName());
	LOG4CXX_DEBUG(logger, (char*) "unbinding: " << serviceName);
	this->connection->name_ctx->unbind(*name);
	LOG4CXX_DEBUG(logger, (char*) "unbound: " << serviceName);

	CorbaEndpointQueue* queue = dynamic_cast<CorbaEndpointQueue*> (destination);
	/*
	PortableServer::POA_ptr poa = queue->getPoa();
	*/
	LOG4CXX_DEBUG(logger, (char*) "deleting queue: " << queue);
	delete queue;
	/*
	LOG4CXX_DEBUG(logger, (char*) "destroying poa: " << poa);
	try {
		poa->destroy(true, true);
	} catch (...) {
		LOG4CXX_DEBUG(logger, (char*) "could not destroy poa: " << poa);
	}
	LOG4CXX_DEBUG(logger, (char*) "destroyed poa: " << poa);
	poa = NULL;
	*/
	LOG4CXX_DEBUG(logger, (char*) "destination destroyed: " << serviceName);
	free(serviceName);
}

File: jatmibroker-xatmi/src/main/java/org/jboss/blacktie/jatmibroker/core/corba/ConnectionImpl.java
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2008, Red Hat, Inc., and others contributors as indicated
 * by the @authors tag. All rights reserved.
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
package org.jboss.blacktie.jatmibroker.core.corba;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.jboss.blacktie.jatmibroker.JAtmiBrokerException;
import org.jboss.blacktie.jatmibroker.core.Transport;
import org.jboss.blacktie.jatmibroker.core.OrbManagement;
import org.jboss.blacktie.jatmibroker.core.Receiver;
import org.jboss.blacktie.jatmibroker.core.Sender;
import org.omg.CORBA.ORBPackage.InvalidName;
import org.omg.CosNaming.NamingContextPackage.AlreadyBound;
import org.omg.CosNaming.NamingContextPackage.CannotProceed;
import org.omg.CosNaming.NamingContextPackage.NotFound;
import org.omg.CosTransactions.TransactionFactory;
import org.omg.CosTransactions.TransactionFactoryHelper;
import org.omg.PortableServer.POAManagerPackage.AdapterInactive;

public class ConnectionImpl implements Runnable, Transport {

	private static final Logger log = LogManager
			.getLogger(ConnectionImpl.class);
	private org.omg.CORBA.Object serverObject;

	private Thread callbackThread;

	private Map<String, SenderImpl> senders = new HashMap<String, SenderImpl>();
	private Map<java.lang.Integer, ReceiverImpl> temporaryQueues = new HashMap<java.lang.Integer, ReceiverImpl>();
	private OrbManagement orbManagement;

	ConnectionImpl(OrbManagement orbManagement, String serverName,
			String password) throws InvalidName, NotFound, CannotProceed,
			org.omg.CosNaming.NamingContextPackage.InvalidName,
			AdapterInactive, AlreadyBound {
		this.orbManagement = orbManagement;

		callbackThread = new Thread(this);
		callbackThread.setDaemon(true);
		callbackThread.start();
	}

	public void close() {
		Iterator<SenderImpl> iterator = senders.values().iterator();
		while (iterator.hasNext()) {
			iterator.next().close();
		}
		Iterator<ReceiverImpl> receivers = temporaryQueues.values().iterator();
		while (iterator.hasNext()) {
			receivers.next().close();
		}
		orbManagement.close();
	}

	public void run() {
		orbManagement.getOrb().run();
	}

	public TransactionFactory getTransactionFactory(
			String transactionManagerServiceName) throws NotFound,
			CannotProceed, org.omg.CosNaming.NamingContextPackage.InvalidName {
		org.omg.CORBA.Object aObject = orbManagement.getNamingContextExt()
				.resolve_str(transactionManagerServiceName);
		return TransactionFactoryHelper.narrow(aObject);
	}

	public Sender getSender(String serviceName) throws JAtmiBrokerException {
		SenderImpl sender = senders.get(serviceName);
		if (sender == null) {
			try {
				org.omg.CORBA.Object serviceFactoryObject = orbManagement
						.getNamingContext().resolve(
								orbManagement.getNamingContextExt().to_name(
										serviceName));
				sender = new SenderImpl(serviceFactoryObject, serviceName);
				senders.put(serviceName, sender);
			} catch (Throwable t) {
				throw new JAtmiBrokerException(
						"Could not load service manager proxy for: "
								+ serviceName, t);
			}
		}
		return sender;
	}

	public Sender createSender(String callback_ior) {
		org.omg.CORBA.Object serviceFactoryObject = orbManagement.getOrb()
				.string_to_object(callback_ior);
		SenderImpl sender = new SenderImpl(serviceFactoryObject, callback_ior);
		return sender;
	}

	public Receiver createReceiver(String serviceName)
			throws JAtmiBrokerException {
		log.debug("createClientCallback create client callback ");
		ReceiverImpl receiver = new ReceiverImpl(orbManagement, serviceName);
		return receiver;
	}

	public Receiver getReceiver(int id) throws JAtmiBrokerException {
		ReceiverImpl receiver = temporaryQueues.get(id);
		if (receiver == null) {
			try {
				log.debug("createClientCallback create client callback ");
				receiver = new ReceiverImpl(orbManagement.getOrb(),
						orbManagement.getRootPoa(), "TODO");
				temporaryQueues.put(id, receiver);
			} catch (Throwable t) {
				throw new JAtmiBrokerException(
						"Could not create a temporary queue", t);
			}
		}
		return receiver;
	}

	public OrbManagement getOrbManagement() {
		return orbManagement;
	}
}
